pkg/broker/instance_operations.go

```diff
@@ -37,21 +40,78 @@ func (b Broker) Provision(ctx context.Context, instanceID string, details broker
 		return
 	}
 
-	// Create a new Atlas cluster from the generated definition
-	resultingCluster, err := b.atlas.CreateCluster(*cluster)
+	// If cluster doesn't exist create it, otherwise compare them
+	resultingCluster, err := b.atlas.GetCluster(cluster.Name)
 	if err != nil {
-		b.logger.Errorw("Failed to create Atlas cluster", "error", err, "cluster", cluster)
-		err = atlasToAPIError(err)
-		return
+		// Create a new Atlas cluster from the generated definition
+		resultingCluster, err = b.atlas.CreateCluster(*cluster)
+		if err != nil {
+			b.logger.Errorw("Failed to create Atlas cluster", "error", err, "cluster", cluster)
+			err = atlasToAPIError(err)
+			return
+		}
+
+		b.logger.Infow("Successfully started Atlas creation process", "instance_id", instanceID, "cluster", resultingCluster)
+
+		return brokerapi.ProvisionedServiceSpec{
+			IsAsync:       true,
+			OperationData: OperationProvision,
+			DashboardURL:  b.atlas.GetDashboardURL(resultingCluster.Name),
+		}, nil
 	}
 
-	b.logger.Infow("Successfully started Atlas creation process", "instance_id", instanceID, "cluster", resultingCluster)
+	err = CompareAndReturnAppropiateResponseCode(resultingCluster, cluster)
+	return
+}
 
-	return brokerapi.ProvisionedServiceSpec{
-		IsAsync:       true,
-		OperationData: OperationProvision,
-		DashboardURL:  b.atlas.GetDashboardURL(resultingCluster.Name),
-	}, nil
+// CompareAndReturnAppropiateResponseCode converts structs to maps and afterwards returns the appropiate response code
+func CompareAndReturnAppropiateResponseCode(resultingCluster *atlas.Cluster, cluster *atlas.Cluster) error {
+	//Convert structs to maps
+	var remoteClusterInterface map[string]interface{}
+	var localClusterInterface map[string]interface{}
+	inrec, _ := json.Marshal(resultingCluster)
+	json.Unmarshal(inrec, &remoteClusterInterface)
+
+	inrec, _ = json.Marshal(cluster)
+	json.Unmarshal(inrec, &localClusterInterface)
+
+	if compareHelper(remoteClusterInterface, localClusterInterface) {
+		return nil
+	}
+	return apiresponses.NewFailureResponse(errors.New("There IDs are equal but differ in their attributes"), http.StatusConflict, "")
+}
+
+func compareHelper(remoteClusterInterface map[string]interface{}, localClusterInterface map[string]interface{}) bool {
+	if reflect.ValueOf(localClusterInterface).Kind() == reflect.Map {
+		for k, v := range localClusterInterface {
+			if reflect.ValueOf(v).Kind() == reflect.Map && len(v.(map[string]interface{})) != 0 {
+				equal := compareHelper(remoteClusterInterface[k].(map[string]interface{}), v.(map[string]interface{}))
```

Comment:
if the `k` key does not exist in the `remoteClusterInterface` map this will panic. Should we check first?

Reply:
I assumed that the response from atlas will always contain the key as described by the docs, but from a robust perspective adding your suggestion is safe and guards the code from breaking, because the user might add a wrong key too from the client-side. Better safe than sorry.